### PR TITLE
Navigator/FlightTaskAuto yaw handling improvements/simplifications

### PR DESCRIPTION
**Pull request by sfuhrer**
### Solved Problem
Fixes https://github.com/PX4/PX4-Autopilot/pull/22413
Alternative to https://github.com/PX4/PX4-Autopilot/pull/22455

Beside it's also a proposal to make the logic around yaw setpoints in the Navigator vs FlightTasks simpler and cleaner.

### Solution
Strictly use the following concept: If Navigator publishes a finite yaw setpoint (`position_setpoint_tripplet.current.yaw`), then disable weather vane. This should be used if waypoints have an associated yaw setpoint (eg landings, VTOL transitions or waypoint mission items with the yaw field set). Otherwise leave authority over yaw to the FlightTask - it can then set it e.g. to direction of the next waypoint, to current yaw, or point the vehicle into the wind with the use of weather vane. 

That allows us to remove the `disable_weather_vane` field from the position_setpoint.

Additionally I propose to thin out the position_setpoint topic and remove unnecessary fields: 
- `yaw_valid` (yaw is declared valid when `yaw` field is finite) --> we should init the yaw field to NAN ideally
- `yawspeed` (doesn't seem to be used)
- `yawspeed_valid`

There's also some other small cleanup proposal, check commits. 

### Changelog Entry
For release notes:
```
Bugfix: FlightTaskAuto: disable wather vane if position setpoint contains yaw setpoint.
```

### Test coverage
SITL tested for MC and VTOL. 
MC test:
- Start the windy gazebo environment: `make px4_sitl gazebo_iris__windy`
- Enable parameter `WV_EN`
- Takeoff and RTL --> on current main it is stuck, doesn't RTL. With the fix here sets the yaw to the home_position.yaw, disables weather vane because the yaw field is finite, aligns yaw and descends. 

### Context
We should remove the yaw field of the Takeoff mission item in QGC, it has no effect (already previously to this PR).  Conceptionally I don't see the need of a yaw attribute to a MC takeoff, and wouldn't know what to expect (yaw once takeoff altitude achieved? Or already when climbing?)
image

### CODE REVIEW

File: src/modules/navigator/navigator_main.cpp [hunk 1]
```diff
@@ -1040,7 +1034,6 @@ void Navigator::geofence_breach_check()
 
 				rep->current.timestamp = hrt_absolute_time();
 				rep->current.yaw = get_local_position()->heading;
```
Comment (MaEtUgR) [inline src/modules/navigator/navigator_main.cpp:1036]:
```suggestion
				rep->current.yaw = NAN;
```

**Review by MaEtUgR — COMMENTED** [review]
Really nice cleanup. Instead of hardcoding a heading for each waypoint from the current vehicle state (ignoring resets) let the trajectory handle the yaw if there's no specific absolute desired heading 👍 

Let's make sure it gets reset to NAN in all those cases otherwise it takes the previous heading.

File: src/modules/navigator/mission_block.cpp [hunk 2]
```diff
@@ -868,7 +863,7 @@ MissionBlock::set_vtol_transition_item(struct mission_item_s *item, const uint8_
 	// Keep yaw from previous mission item if valid, as that is containing the transition heading.
 	// If not valid use current yaw as yaw setpoint
 	if (!PX4_ISFINITE(item->yaw)) {
-		item->yaw = _navigator->get_local_position()->heading; // ideally that would be course and not heading
+		item->yaw = NAN; // ideally that would be course and not heading
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/navigator/mission_block.cpp]: done in 7c9aebe0497f504ae8a7afa1c06d0f578bdc70a8

File: src/modules/navigator/navigator_main.cpp [hunk 3]
```diff
@@ -349,7 +347,6 @@ void Navigator::run()
 						    && (get_position_setpoint_triplet()->current.type != position_setpoint_s::SETPOINT_TYPE_TAKEOFF)) {
 
 							calculate_breaking_stop(rep->current.lat, rep->current.lon, rep->current.yaw);
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/navigator/navigator_main.cpp]: done in 83f23912bf7fa37859b92f28d66100bca756eb93, and made sure it's set to NAN instead in 81e2662fea71d1f95f113de51c8f381472310a05

File: src/modules/navigator/navigator_main.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/modules/navigator/navigator_main.cpp]: done in 81e2662fea71d1f95f113de51c8f381472310a05

**Comment by sfuhrer** [comment]
@MaEtUgR addressed your comments and additionally fixed the init of the triplet.yaw to NAN.

**Comment by TedObrien** [comment]
Hi @sfuhrer, thanks for taking the time to look at this issue, I just tested it and it solves the problem I was having (#22413). I did have one thought regarding this solution:

If weather vane is disabled when a finite yaw setpoint is given, it will always be disabled during RTL, as there is currently always a yaw setpoint given. This may be a problem in the case of an asymetric airframe or a vehicle with a lifting surface, where it may be desirable to point into the direction of the wind the entire time. A parameter to not provide a yaw setpoint (enable weather vane) during RTL might be useful in some use cases. Hope this makes sence.

**Comment by sfuhrer** [comment]
> If weather vane is disabled when a finite yaw setpoint is given, it will always be disabled during RTL, as there is currently always a yaw setpoint given. 

Valid input! I already kicked off the discussion about removing the special yaw handling for RTL, see https://github.com/PX4/PX4-Autopilot/pull/18170#issuecomment-1852019861. Though maybe we should remove it, but instead add a new option for RTL_HDG_MD, "no yaw setpoint"? Then the FlightTask would decide on the yaw setpoint, and if WV is enabled it would keep working.

**Comment by TedObrien** [comment]
A new "no yaw setpoint" for RTL_HDG_MD  sounds good to me. 
As for removing the current yaw handling, using the home heading for RTL can be useful as its much easier/safer take over control during take off and landing when the vehicle is facing away from you.

**Review by MaEtUgR — APPROVED** [review]
Thanks, everything I saw in my review was addressed.
It's a good step forward allowing to have the controller handle yaw if it's not explicitly set to a certain angle.

**Comment by sfuhrer** [comment]
> A new "no yaw setpoint" option for RTL_HDG_MD sounds good to me.

@TedObrien could you take over the implementation?

**Comment by TedObrien** [comment]
@sfuhrer Happy to give it a go, should have some time in the new year.

**Pull request merged**